### PR TITLE
fix: do not escape underscore inside of words

### PR DESCRIPTION
**Pull request by nikku**
### Initial checklist

* [x] I read the support docs 
* [x] I read the contributing guide 
* [x] I agree to follow the code of conduct 
* [x] I searched issues and discussions and couldn’t find anything or linked relevant results below 
* [x] I made sure the docs are up to date
* [x] I included tests (or that’s not needed)

### Description of changes

As commented in https://github.com/remarkjs/remark/issues/1455#issuecomment-3724812006 underscores inside of words are fine in commonmark, i.e. `HELLO_WORLD` or `https://raw/links_with_underscore` should not be escaped.

This PR makes escape rules for `_` more specific to prevent unnecessary escaping inside of words such as HELLO_WORLD or urls (https://some/web_site).

Related to https://github.com/remarkjs/remark/issues/1455

### CODE REVIEW

**Comment by nikku** [comment]
I see `npm test` reformat a bunch of source files that I did not touch. I suggest to run `npm test` and apply the auto formatting for these files separately.

**Comment by wooorm** [comment]
Thanks but I’d like to see some more round tripping tests! Whitespace is not enough. Punctuation affects things too. See CM for the complexity.

Things are this way for a reason. It may be possible to improve things. But it’s not going to very be simple.

thanks!

**Comment by nikku** [comment]
I don't know what `CM` refers to, can you elaborate?

Regarding roundtrip tests, I [added](https://github.com/syntax-tree/mdast-util-to-markdown/pull/72/commits/36591c02dd7a1bb224474f147897cd76701dab0b) [some](https://github.com/syntax-tree/mdast-util-to-markdown/pull/72/commits/27525c142c9286b731badafde0c917dfd13d2a58) and would apprechiate your feedback if you see any additional scenarios missing. The change does not break with any of the existing tests, which is encouraging. 

I also ran it against some of our projects, and could not see it break things. However, I found that remark is escaping in other places, to mention two: 

| Input | Output |
|:---|:---|
| `a@b` | `a\@b` |
| `[...]` | `\[...\]` |

This leads me to the question: What is the goal of this library, and the formating that `remark-cli` offers?

* to create as safe as possible markdown, from any AST (or valid ASTs)
* to produce markdown as per a certain standard
* to be able to repeatedly roundtrip things (`markdown START -> ast -> markdown -> ast -> markdown END`, where `markdown START` and `markdown END` are the same thing?
* ...

**Comment by ChristianMurphy** [comment]
> I don't know what CM refers to, can you elaborate?

CommonMark https://commonmark.org/

> feedback if you see any additional scenarios missing

My mind goes to making sure that it doesn't mess with emphasis encoding

https://github.com/nikku/mdast-util-to-markdown/blob/27525c142c9286b731badafde0c917dfd13d2a58/test/index.js#L4834-L4859

And handle some nesting conditions https://github.com/syntax-tree/mdast-util-to-markdown/issues/12 (though not necessarily this exact one since its already broken 😅 )

> to create as safe as possible markdown, from any AST (or valid ASTs)

For valid ASTs

> to produce markdown as per a certain standard

Yes the CommonMark standard

> to be able to repeatedly roundtrip things (markdown START -> ast -> markdown -> ast -> markdown END, where markdown START and markdown END are the same thing?

Close, to handle valid ASTs and keep with the standard.
And to keep structural integrity AST roundtrip ends with same AST structure.

A test I've run in the past is a fuzzer pass.
Usually jsfuzz https://github.com/micromark/micromark/blob/main/test/fuzz.cjs

with a setup like

```
+-------------------+      parse      +-------------------+      stringify   +-------------------+      parse      +-------------------+
|   text (markdown) |  -------------> |     mdast (AST)    |  -------------> |   text (markdown) |  -------------> |     mdast (AST)    |
+-------------------+                 +-------------------+                  +-------------------+                 +-------------------+

INVARIANT (structural integrity):
  For every node position in the tree, the node "type" must match.
  (Same mdast structure by node.type across the roundtrip.)
```

> markdown START and markdown END are the same thing?

the remark ecosystem keeps structural information, not exact text information.
So the raw text can, and will shift.
But the output document should remain the same structurally.

**Comment by nikku** [comment]
> My mind goes to making sure that it doesn't mess with emphasis encoding

Cf. https://github.com/syntax-tree/mdast-util-to-markdown/pull/72/commits/30e46e7d889318aed3a36e88418416c63a7e9307. From what it looks it does not mess with the existing behavior.

**Comment by nikku** [comment]
@ChristianMurphy and because you mention edge cases - I'm sure there is some. But you have solid test coverage in place, that ensure that the _standard_, happy path is working. It is.

File: test/index.js [hunk 1]
```diff
@@ -4789,7 +4870,15 @@ a ___ is this emphasis? ___
...
-a _\\__ is this emphasis? _\\__`
+a _\\__ is this emphasis? _\\__
+
+a _ is_this emphasis? _
+a __ is_this emphasis? __
+a ___ is_this emphasis? ___
+a _\\_ is_this emphasis? _\\_
+a \\__ is_this emphasis? \\__
+a __\\_ is_this emphasis? __\\_
+a _\\__ is_this emphasis? _\\__`
```
Comment (wooorm) [inline test/index.js:4881]: The needed tests are as follows:

* a matrix of whitespace, punctuation, and rest (such as ` `, `.`, `x`)
* all 4 sides (before open, after open, before close, after close)
* markers (`_` and `*`)
* marker count (1, 2)
* nesting (1, 2, 3 times is probably enough)

I think that adds up to `3*4*2*2*3 == 144`. You’d end up with say `a.*.b.*.c`. Where the `.`s are the surrounding characters, and `*` the markers.

You should probably generate those cases! Like what @ChristianMurphy mentioned above, it is important that they are structurally markdown->ast->markdown->ast equal, but the 2 markdown strings do not have to be equal

**Comment by tats-u** [comment]
You can safely keep `_` as is without additional conditions iff neither of the adjacent characters are [Unicode whitespace](https://spec.commonmark.org/0.31.2/#unicode-whitespace-character) or [Unicode punctuation](https://spec.commonmark.org/0.31.2/#unicode-punctuation-character) characters, or both of them are Unicode whitespace characters.

`a(_)b(_)c` → `a(<em>)b(</em>)c` ([micromark/markdown-it/marked](https://tats-u.github.io/markdown-cjk-friendly/?src=a%28_%29b%28_%29c&gfm=1&engine=micromark&ver=2.0.1) | [commonmark.js](https://spec.commonmark.org/dingus/?text=a(_)b(_)c)) because parenthesis are Unicode punctuation characters

Note: the definition of the Unicode punctuation character is slightly different between GitHub's cmark-gfm (https://github.github.com/gfm/#punctuation-character) and the other CommonMark-compliant parsers. GitHub doesn't treat characters whose [general category](https://unicode.org/glossary/#general_category) is `S` as Unicode punctuation characters. (GitHub treats "punctuation" as "Unicode punctuation character" (P) + "ASCII punctuation character" ((P or S) + ASCII), but the latest CommonMark treats "punctuation" as Unicode punctuation character (P or S))

**Comment by nikku** [comment]
I'll not be able to follow-up on this work. If you deem actual output (and readability) important, then let's get these improvements in, somehow - not violating the integrity that you're talking about (which I honestly cannot evaluate). 

As an end user, I care about readable Markdown :slightly_smiling_face:.
